test-utils/state-viewer/src/state_dump.rs

```diff
@@ -42,66 +42,99 @@ pub fn state_dump(
-    // minting tokens every epoch.
-    genesis_config.total_supply = total_supply;
     let shard_config = runtime.get_shard_config(last_block_header.epoch_id()).unwrap();
     genesis_config.shard_layout = shard_config.shard_layout;
     genesis_config.num_block_producer_seats_per_shard =
         shard_config.num_block_producer_seats_per_shard;
     genesis_config.avg_hidden_validator_seats_per_shard =
         shard_config.avg_hidden_validator_seats_per_shard;
-    near_config.genesis = Genesis::new_with_path(genesis_config, records_path.to_path_buf());
     // Record only the filename of the records file.
     // Otherwise the absolute path is stored making it impossible to copy the dumped state to actually use it.
-    near_config.config.genesis_records_file =
-        Some(records_path.file_name().unwrap().to_str().unwrap().to_string());
+    match records_path {
+        Some(records_path) => {
+            let mut records_path_dir = records_path.to_path_buf();
+            records_path_dir.pop();
+            fs::create_dir_all(&records_path_dir).unwrap_or_else(|_| {
+                panic!("Failed to create directory {}", records_path_dir.display())
+            });
+            let records_file = File::create(&records_path).unwrap();
+            let mut ser = serde_json::Serializer::new(records_file);
+            let mut seq = ser.serialize_seq(None).unwrap();
+            let total_supply =
+                get_records(runtime, state_roots, last_block_header, &validators, |sr| {
+                    seq.serialize_element(&sr).unwrap()
+                });
+            seq.end().unwrap();
+            // `total_supply` is expected to change due to the natural processes of burning tokens and
+            // minting tokens every epoch.
+            genesis_config.total_supply = total_supply;
+            near_config.genesis =
+                Genesis::new_with_path(genesis_config, records_path.to_path_buf());
+            near_config.config.genesis_records_file =
+                Some(records_path.file_name().unwrap().to_str().unwrap().to_string());
+        }
+        None => {
+            let mut records: Vec<StateRecord> = vec![];
+            let total_supply =
+                get_records(runtime, state_roots, last_block_header, &validators, |sr| {
+                    records.push(sr)
+                });
+            // `total_supply` is expected to change due to the natural processes of burning tokens and
+            // minting tokens every epoch.
+            genesis_config.total_supply = total_supply;
+            near_config.genesis = Genesis::new(genesis_config, records.into());
+        }
+    }
     near_config
 }
 
+fn get_records<'a, F, X>(
+    runtime: NightshadeRuntime,
+    state_roots: Vec<StateRoot>,
+    last_block_header: BlockHeader,
+    validators: &'a HashMap<AccountId, (X, Balance)>,
+    mut callback: F,
+) -> Balance
+where
+    F: FnMut(StateRecord),
```

Comment:
It feels a bit weird that we need this to be generic. As a rule of thumb, I try to avoid generics in Rust, to avoid code bloat and longer compile times. Will try to take a closer look tomorrow whether this can be simplifield.